apps/meteor/app/livechat/server/api/lib/users.ts

```diff
@@ -10,20 +11,62 @@ import { escapeRegExp } from '@rocket.chat/string-helpers';
 async function findUsers({
 	role,
 	text,
+	onlyAvailable = false,
+	excludeId,
+	showIdleAgents = true,
 	pagination: { offset, count, sort },
 }: {
 	role: IRole['_id'];
 	text?: string;
+	onlyAvailable?: boolean;
+	excludeId?: string;
+	showIdleAgents?: boolean;
 	pagination: { offset: number; count: number; sort: any };
 }): Promise<{ users: ILivechatAgent[]; count: number; offset: number; total: number }> {
-	const query = {};
+	const query: FilterOperators<ILivechatAgent> = {};
 	if (text) {
 		const filterReg = new RegExp(escapeRegExp(text), 'i');
 		Object.assign(query, {
 			$or: [{ username: filterReg }, { name: filterReg }, { 'emails.address': filterReg }],
 		});
 	}
 
+	if (onlyAvailable) {
+		Object.assign(query, {
+			statusLivechat: 'available',
+		});
+	}
+
+	if (excludeId) {
+		Object.assign(query, {
+			_id: { $ne: excludeId },
+		});
+	}
+
+	if (!showIdleAgents) {
+		const oldOr = query.$or;
+
+		const newOr = [{ status: { $exists: true, $ne: 'offline' }, roles: { $ne: 'bot' } }, { roles: 'bot' }];
+
+		if (oldOr) {
```

Comment:
if I'm understanding the idea correctly, I would suggest doing something like this to avoid messing up with a `query` object

```js
const conditions = [];

if (text) {
  conditions.push({ $or: [{ username: filterReg }, { name: filterReg }, { 'emails.address': filterReg }] });
}

if (excludeId) {
  conditions.push({ _id: { $ne: excludeId } });
}

if (!showIdleAgents) {
  conditions.push({ $or: [{ status: { $exists: true, $ne: 'offline' }, roles: { $ne: 'bot' } }, { roles: 'bot' }] });
}

const query = { $and: conditions };
```

I THINK the result is the same.. wdyt?

Reply:
Let me do something similar